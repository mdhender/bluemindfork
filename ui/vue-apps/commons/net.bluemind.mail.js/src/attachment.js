import { MimeType } from "@bluemind/email";
import { inject } from "@bluemind/inject";

export function create(part, status) {
    const progress = status === AttachmentStatus.NOT_LOADED ? { loaded: 0, total: 100 } : { loaded: 100, total: 100 };
    if (!part.fileName) {
        part.fileName = inject("i18n").t("mail.attachment.untitled", { mimeType: part.mime });
    }
    if (part.mime === "application/octet-stream" && part.fileName && getTypeFromExtension(part.fileName)) {
        part.mime = getTypeFromExtension(part.fileName);
    }
    const headers = part.headers ? [...getAttachmentHeaders(part), ...part.headers] : getAttachmentHeaders(part);

    let attachment = {
        dispositionType: "ATTACHMENT",
        progress,
        status,
        type: "default",
        extra: {},
        ...part,
        headers
    };
    return attachment;
}

function getTypeFromExtension(filename) {
    return MimeType.getFromFilename(filename);
}

export const AttachmentStatus = {
    ONLY_LOCAL: "ONLY_LOCAL",
    NOT_LOADED: "NOT-LOADED",
    UPLOADED: "UPLOADED",
    ERROR: "ERROR"
};

export function getAttachmentHeaders({ fileName, size }) {
    return [
        {
            name: "Content-Disposition",
            values: ["attachment;filename=" + fileName + ";size=" + size]
        },
        {
            name: "Content-Transfer-Encoding",
            values: ["base64"]
        }
    ];
}

/**
 * @return true if we consider the given part should be attached,
 *         i.e.: not shown in the message body, false otherwise
 */
export function isAttachment(part) {
    return (part.dispositionType && part.dispositionType !== "INLINE") || part.mime === "application/octet-stream";
}

export default {
    create,
    AttachmentStatus,
    getAttachmentHeaders,
    isAttachment
};
