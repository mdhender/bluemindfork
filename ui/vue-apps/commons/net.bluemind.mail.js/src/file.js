import { isViewable } from "./part";

export const FileStatus = {
    ONLY_LOCAL: "ONLY_LOCAL",
    NOT_LOADED: "NOT-LOADED",
    UPLOADED: "UPLOADED",
    ERROR: "ERROR",
    INVALID: "INVALID"
};

const LARGE_FILE_SIZE = 100 * 1024 * 1024;
const VERY_LARGE_FILE_SIZE = 500 * 1024 * 1024;

function isUploading({ status }) {
    return ![FileStatus.UPLOADED, FileStatus.ERROR, FileStatus.INVALID].includes(status);
}

function isLarge({ size }) {
    return size > LARGE_FILE_SIZE;
}

function isAllowedToPreview(file) {
    return isViewable(file) && file.size && !isUploading(file) && !isLarge(file);
}

function hasRemoteContent({ url }) {
    return url && url.startsWith("http");
}

export default {
    FileStatus,
    hasRemoteContent,
    isAllowedToPreview,
    isLarge,
    isUploading,
    LARGE_FILE_SIZE,
    VERY_LARGE_FILE_SIZE
};
