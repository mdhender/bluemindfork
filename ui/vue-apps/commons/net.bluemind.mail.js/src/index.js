export { default as attachmentUtils } from "./attachment";
export { default as conversationUtils } from "./conversations";
export { default as draftUtils } from "./draft";
export { default as fileUtils } from "./file";
export { default as folderUtils } from "./folder";
export { default as loadingStatusUtils } from "./loading-status";
export { default as mailboxUtils } from "./mailbox";
export { default as mailTipUtils } from "./mailTip";
export { default as messageUtils } from "./message";
export { default as partUtils } from "./part";
export { default as signatureUtils } from "./signature";
