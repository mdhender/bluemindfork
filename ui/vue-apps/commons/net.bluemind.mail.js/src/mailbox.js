import { LoadingStatus } from "./loading-status";

export const MailboxType = {
    MAILSHARE: "mailshares",
    USER: "users"
};

export function create({ owner, dn, address, type }) {
    switch (type) {
        case MailboxType.USER:
            return createUserMailbox({ owner, dn, address });
        case MailboxType.MAILSHARE:
            return createSharedMailbox({ owner, dn, address });
    }
}

function createUserMailbox({ owner, dn, address }) {
    return {
        ...createBaseMailbox({ owner, name: address, dn, address }),
        type: MailboxType.USER,
        remoteRef: {
            uid: "user." + owner
        },
        key: "user." + owner,
        root: ""
    };
}

function createSharedMailbox({ owner, dn, address }) {
    return {
        ...createBaseMailbox({ owner, name: dn, dn, address }),
        type: MailboxType.MAILSHARE,
        remoteRef: {
            uid: owner
        },
        key: owner,
        root: dn
    };
}

function createBaseMailbox({ owner, name, dn, address }) {
    return {
        dn,
        address,
        owner,
        name,
        loading: LoadingStatus.NOT_LOADED,
        writable: true
    };
}

export default {
    create,
    MailboxType
};
