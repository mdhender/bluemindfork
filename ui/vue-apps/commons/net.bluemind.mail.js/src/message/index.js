import cloneDeep from "lodash.clonedeep";
import escapeRegExp from "lodash.escaperegexp";
import merge from "lodash.merge";
import pick from "lodash.pick";
import random from "lodash.random";

import { Flag } from "@bluemind/email";

import { LoadingStatus } from "../loading-status";
import EmlParser from "./EmlParser";
import MessageAdaptor from "./MessageAdaptor";

export function createOnlyMetadata({ internalId, folder: { key, uid }, conversationRef, date }) {
    return {
        key: internalId >= 0 && key ? messageKey(internalId, key) : null,
        folderRef: { key, uid },
        remoteRef: { internalId },
        conversationRef,
        status: MessageStatus.IDLE,
        loading: LoadingStatus.NOT_LOADED,
        date
    };
}

export function messageKey(id, folderKey) {
    const string = folderKey + "/" + id;
    let hash = 0;
    for (let i = 0; i < string.length; i++) {
        hash = ((hash << 5) - hash + string.charCodeAt(i)) & 0xffffffff;
    }

    return hash;
}

export function createWithMetadata(metadata) {
    const messageMetadata = createOnlyMetadata(metadata);
    const emptyMessage = create();
    return merge(emptyMessage, messageMetadata);
}

export function create() {
    const emptyData = {
        remoteRef: { imapUid: null },
        flags: [],
        date: null,
        headers: [],
        subject: "",
        preview: "",
        size: 0,
        composing: false,
        version: null,

        // sender & recipients
        from: {
            address: "",
            dn: ""
        },
        to: [],
        cc: [],
        bcc: [],

        // used only by reply / forward
        messageId: "",
        conversationId: undefined,

        // parts
        hasAttachment: false,
        hasICS: false,

        eventInfo: {
            isCounterEvent: false,
            icsUid: "",
            needsReply: false
        },

        attachments: [],
        inlinePartsByCapabilities: [],
        reports: []
    };
    return merge(createOnlyMetadata({ folder: {} }), emptyData);
}

export function partialCopy(message, properties = []) {
    return cloneDeep(pick(message, properties.concat("key", "folderRef", "status", "remoteRef")));
}

export const MessageStatus = {
    IDLE: "IDLE",
    NEW: "NEW",
    SAVING: "SAVING",
    SAVE_ERROR: "SAVE_ERROR",
    SENDING: "SENDING",
    SENT: "SENT",
    INVALID: "INVALID"
};

export const MessageCreationModes = {
    NEW: "NEW",
    REPLY: "REPLY",
    REPLY_ALL: "REPLY-ALL",
    FORWARD: "FORWARD",
    EDIT_AS_NEW: "EDIT_AS_NEW",
    EDIT: "EDIT",
    FORWARD_AS_EML: "FORWARD_AS_EML"
};

export const MessageHeader = {
    MAIL_FOLLOWUP_TO: "Mail-Followup-To",
    MAIL_REPLY_TO: "Mail-Reply-To",
    MESSAGE_ID: "Message-ID",
    REPLY_TO: "Reply-To",
    IN_REPLY_TO: "In-Reply-To",
    REFERENCES: "References",
    DELIVERED_TO: "Delivered-To",
    DISPOSITION_NOTIFICATION_TO: "Disposition-Notification-To",
    ORIGINAL_RECIPIENT: "Original-Recipient",

    X_LOOP: "X-Loop",
    X_ORIGINAL_TO: "X-Original-To",

    X_BM_DRAFT_INFO: "X-Bm-Draft-Info",
    X_BM_DRAFT_REFRESH_DATE: "X-Bm-Draft-Refresh-Date",
    X_BM_EVENT_COUNTERED: "X-BM-Event-Countered",
    X_BM_EVENT: "X-BM-Event",
    X_BM_RESOURCEBOOKING: "X-BM-ResourceBooking",
    X_BM_REWRITE: "X-BM-Rewrite",
    X_BM_SENT_FOLDER: "X-BM-Sent-Folder"
};

export const MessageReplyAttributeSeparator = "data-bm-reply-separator";
export const MessageForwardAttributeSeparator = "data-bm-forward-separator";
export const MessageQuoteMozillaClass = "moz-cite-prefix";
export const MessageQuoteGmailClass = "gmail_quote";
export const MessageQuoteProtonClass = "protonmail_quote";
export const MessageQuoteYahooClass = "yahoo_quoted";
export const MessageQuoteClasses = [
    MessageReplyAttributeSeparator,
    MessageForwardAttributeSeparator,
    MessageQuoteMozillaClass,
    MessageQuoteGmailClass,
    MessageQuoteProtonClass,
    MessageQuoteYahooClass
];
export const MessageQuoteOutlookId = "divRplyFwdMsg";

export function isUnread(message) {
    return message.loading === LoadingStatus.LOADED && !message.flags.includes(Flag.SEEN);
}

export function isFlagged(message) {
    return message.loading === LoadingStatus.LOADED && message.flags.includes(Flag.FLAGGED);
}

/** Extract multi-valued / whitespace separated values from given header. */
export function extractHeaderValues(message, headerName) {
    const header = message.headers.find(h => h.name.toUpperCase() === headerName.toUpperCase());
    return header && header.values && header.values.length
        ? header.values.reduce((a, b) => (a.length ? a + " " + b : b), "").split(/\s+/)
        : undefined;
}

/**
 * Try to detect a message is a Forward.
 * @see https://stackoverflow.com/questions/4735293/forwarded-email-detection
 */
export function isForward(message) {
    let hasAForwardFriendlyHeader = false;
    let hasBmDraftHeaderWithFwdType = false;

    message.headers?.forEach(header => {
        // MUAs should add one of these headers
        if (
            [
                MessageHeader.MAIL_FOLLOWUP_TO.toUpperCase(),
                MessageHeader.DELIVERED_TO.toUpperCase(),
                MessageHeader.X_LOOP.toUpperCase(),
                MessageHeader.X_ORIGINAL_TO.toUpperCase()
            ].includes(header.name.toUpperCase())
        ) {
            hasAForwardFriendlyHeader = true;
        }

        // BM specific
        if (
            header.name.toUpperCase() === MessageHeader.X_BM_DRAFT_INFO.toUpperCase() &&
            header.values?.some(v => JSON.parse(v).type === MessageCreationModes.FORWARD)
        ) {
            hasBmDraftHeaderWithFwdType = true;
        }
    });

    // MUAs tend to add the 'Fwd' token in the subject (which may be rewritten by the user)
    const subjectStartsWithFwd = /^\[*Fwd?:/i.test(message.subject);
    const subjectEndsWithFwd = /\(fwd?\)$/i.test(message.subject);
    const hasAForwardFriendlySubject = subjectStartsWithFwd || subjectEndsWithFwd;

    const shouldBeForward = hasBmDraftHeaderWithFwdType || hasAForwardFriendlyHeader || hasAForwardFriendlySubject;

    return shouldBeForward;
}

export function isReply(message) {
    return message.headers?.some(header =>
        [MessageHeader.IN_REPLY_TO.toUpperCase(), MessageHeader.REFERENCES.toUpperCase()].includes(
            header.name.toUpperCase()
        )
    );
}

/**
 * Create a portable name for the EML file of the given message.
 * @see  https://www.mtu.edu/umc/services/websites/writing/characters-avoid/
 * @param {Object} message the message used to generate an EML file name
 * @param {String} fallBackName a fall-back name if the use of message has failed (ex: empty subject)
 * @returns a portable EML file name
 */
export function createEmlName(message, fallBackName) {
    const maxChars = 31;
    const extension = ".eml";
    const illegalChars = "#<>$+%!`&*'|{}?\"=/\\:@";
    const illegalCharsRegex = new RegExp(`[${escapeRegExp(illegalChars)}\\s]+`, "g");
    let subject = message.subject?.trim() || fallBackName;
    const replacementChar = "-";
    subject = subject.replaceAll(illegalCharsRegex, replacementChar);
    if (subject.length > maxChars - extension.length) {
        subject = subject.substring(0, maxChars - extension.length);
    }
    return `${subject}${extension}`;
}

export function findDispositionNotificationHeaderIndex(headers) {
    return headers.findIndex(
        header =>
            new RegExp(MessageHeader.DISPOSITION_NOTIFICATION_TO, "i").test(header.name) &&
            header.values?.filter(Boolean)?.length
    );
}

export function setDispositionNotificationHeader(headers, from) {
    const index = findDispositionNotificationHeaderIndex(headers);
    let header;
    if (index >= 0) {
        header = headers[index];
    } else {
        header = { name: MessageHeader.DISPOSITION_NOTIFICATION_TO };
        headers.push(header);
    }
    header.values = [from.dn ? `${from.dn} <${from.address}>` : from.address];
}

export function removeDispositionNotificationHeader(headers) {
    const index = findDispositionNotificationHeaderIndex(headers);
    if (index >= 0) {
        headers.splice(index, 1);
    }
}

export function generateMessageIDHeader(address) {
    const encodedDate = new Date().getTime().toString(36);
    const encodedRandomNumber = random(Math.pow(2, 64) - 1).toString(36);
    const domain = address.substring(address.indexOf("@") + 1);
    const value = "<" + encodedDate + "." + encodedRandomNumber + "@" + domain + ">";
    return {
        name: MessageHeader.MESSAGE_ID,
        values: [value]
    };
}

export default {
    create,
    createEmlName,
    createOnlyMetadata,
    createWithMetadata,
    extractHeaderValues,
    findDispositionNotificationHeaderIndex,
    generateMessageIDHeader,
    isFlagged,
    isForward,
    isReply,
    isUnread,
    EmlParser,
    MessageAdaptor,
    MessageCreationModes,
    MessageForwardAttributeSeparator,
    MessageHeader,
    messageKey,
    MessageQuoteClasses,
    MessageQuoteGmailClass,
    MessageQuoteMozillaClass,
    MessageQuoteOutlookId,
    MessageQuoteProtonClass,
    MessageQuoteYahooClass,
    MessageReplyAttributeSeparator,
    MessageStatus,
    partialCopy,
    removeDispositionNotificationHeader,
    setDispositionNotificationHeader
};
