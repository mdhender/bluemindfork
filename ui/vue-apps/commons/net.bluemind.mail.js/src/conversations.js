import { MessageCreationModes, messageKey } from "./message";
import { LoadingStatus } from "./loading-status";
import { draftInfoHeader } from "./draft";
import { isDraftFolder } from "./folder";

export function createConversationStub(id, folderRef) {
    return {
        folderRef,
        key: messageKey(id, folderRef.key),
        loading: LoadingStatus.NOT_LOADED,
        messages: [],
        remoteRef: { uid: id }
    };
}

export function firstMessageInConversationFolder(getters, conversations) {
    return conversations
        .map(conversation =>
            getters
                .CONVERSATION_MESSAGE_BY_KEY(conversation.key)
                .find(m => m.folderRef.key === conversation.folderRef.key)
        )
        .filter(Boolean);
}

export function messagesInConversationFolder(getters, conversations) {
    return conversations.flatMap(conversation =>
        getters
            .CONVERSATION_MESSAGE_BY_KEY(conversation.key)
            .filter(m => m.folderRef.key === conversation.folderRef.key)
    );
}

export function firstMessageFolderKey(conversation) {
    return conversation.messages[0].folderRef.key;
}

/**
 * Sort by date, ascending, except for draft which should be just after its related message. */
export function sortConversationMessages(messages, folders) {
    // sort by date
    const sorted = messages.sort((a, b) => a.date - b.date);

    // search for drafts
    const drafts = messages.filter(m => isDraftFolder(folders[m.folderRef.key].path));
    const allowedCreationModes = [
        MessageCreationModes.REPLY_ALL,
        MessageCreationModes.REPLY,
        MessageCreationModes.FORWARD,
        MessageCreationModes.FORWARD_AS_EML
    ];
    drafts.forEach(c => {
        const draftInfo = draftInfoHeader(c);
        if (
            draftInfo &&
            allowedCreationModes.includes(draftInfo.type) &&
            draftInfo.messageInternalId &&
            draftInfo.folderUid
        ) {
            const relatedIndex = sorted.findIndex(
                s => s.remoteRef.internalId === draftInfo.messageInternalId && s.folderRef.uid === draftInfo.folderUid
            );
            const draftIndex = sorted.findIndex(s => s.key === c.key);
            // remove
            const draft = sorted.splice(draftIndex, 1)[0];
            // insert
            sorted.splice(relatedIndex + 1, 0, draft);
        }
    });

    return sorted;
}

export function conversationMustBeRemoved(state, conversation, messages) {
    if (!conversation) {
        return false;
    }
    const keys = new Set(messages.map(message => message.key));
    const folderKey = conversation.folderRef.key;
    return !conversation.messages.some(
        key => !keys.has(key) && state.messages[key] && state.messages[key].folderRef.key === folderKey
    );
}

export function idToUid(value) {
    const bytes = 64n;
    // eslint-disable-next-line no-undef
    const bigIntValue = BigInt(value);
    if (bigIntValue < 0n) {
        return (bigIntValue + (1n << bytes)).toString(16);
    }
    return bigIntValue.toString(16);
}

export default {
    conversationMustBeRemoved,
    createConversationStub,
    firstMessageFolderKey,
    firstMessageInConversationFolder,
    idToUid,
    messagesInConversationFolder,
    sortConversationMessages
};
