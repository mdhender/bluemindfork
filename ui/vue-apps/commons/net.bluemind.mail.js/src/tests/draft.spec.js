import { MockI18NProvider } from "@bluemind/test-utils";
import ServiceLocator from "@bluemind/inject";

import { create, MessageCreationModes, MessageHeader } from "~/src/message";
import {
    computeCcRecipients,
    computeToRecipients,
    computeSubject,
    findIdentityFromMailbox,
    addSeparator,
    computeIdentityForReplyOrForward
} from "~/src/draft";

ServiceLocator.register({ provide: "i18n", factory: () => MockI18NProvider });
const vueI18n = ServiceLocator.getProvider("i18n").get();

const previousMessageFrom = { address: "someone@vm40.net", dn: "Some One" };
const previousMessageTo = [
    { dn: "John Doe", address: "jdoe@vm40.net" },
    { dn: "Toto Matic", address: "tmatic@vm40.net" },
    { dn: "Georges Abitbol", address: "gabitbol@vm40.net" }
];
const previousMessageCc = [
    { dn: "Another John Doe", address: "anotherjdoe@vm40.net" },
    { dn: "Another Toto Matic", address: "anothertmatic@vm40.net" },
    { dn: "Another Georges Abitbol", address: "anothergabitbol@vm40.net" }
];

const previousMessage = {
    ...create(),
    ...{
        date: new Date(),
        from: previousMessageFrom,
        to: previousMessageTo,
        cc: previousMessageCc,
        remoteRef: { internalId: 10 },
        folderRef: { uid: "my-uid" }
    }
};

describe("Compute subject", () => {
    const message = { subject: "TrucTruc" };

    test("computeSubject for Reply", () => {
        checkBuildSubject(message, MessageCreationModes.REPLY, "Re: ");
    });

    test("computeSubject for ReplyAll", () => {
        checkBuildSubject(message, MessageCreationModes.REPLYALL, "Re: ");
    });

    test("computeSubject for Forward", () => {
        checkBuildSubject(message, MessageCreationModes.FORWARD, "Fw: ");
    });
});

function checkBuildSubject(message, creationMode, prefix) {
    const subject = computeSubject(creationMode, message);
    const expectedSubject = prefix + message.subject;
    expect(subject).toEqual(expectedSubject);

    // should not add the prefix again
    const subject2 = computeSubject(creationMode, message);
    expect(subject2).toEqual(expectedSubject);
}

describe("Compute reply / forward separators", () => {
    const messageContent = "messageContent";

    let itemsService = {};
    itemsService.uploadPart = jest.fn().mockReturnValue("2");

    vueI18n.t = jest.fn().mockImplementation((key, params) => {
        if (key === "mail.compose.reply.body") {
            return "On " + params.date + ", " + params.name + " wrote:";
        } else if (key === "mail.compose.forward.body") {
            return "---- Original Message ----";
        } else if (key === "mail.compose.forward.prev.message.info.from") {
            return "From";
        } else if (key === "mail.compose.forward.prev.message.info.to") {
            return "To";
        } else if (key === "mail.compose.forward.prev.message.info.date") {
            return "Date";
        } else if (key === "mail.compose.forward.prev.message.info.subject") {
            return "Subject";
        }
    });

    test("for Reply and ReplyAll with userPrefTextOnly", () => {
        const expectedContent =
            "\n<p>On " + previousMessage.date + ", Some One <someone@vm40.net> wrote:\n</p>\n\n> " + messageContent;

        let contentWithSeparator = addSeparator(
            messageContent,
            previousMessage,
            MessageCreationModes.REPLY,
            true,
            vueI18n
        );
        expect(contentWithSeparator).toEqual(expectedContent);
        contentWithSeparator = addSeparator(
            messageContent,
            previousMessage,
            MessageCreationModes.REPLY_ALL,
            true,
            vueI18n
        );
        expect(contentWithSeparator).toEqual(expectedContent);
    });

    test("for Forward with userPrefTextOnly", async () => {
        const expectedContent =
            '\n<p style="color: purple;">---- Original Message ----\nSubject: ' +
            previousMessage.subject +
            "\nTo: John Doe <jdoe@vm40.net>,Toto Matic <tmatic@vm40.net>,Georges Abitbol <gabitbol@vm40.net>\nDate: " +
            previousMessage.date +
            "\nFrom: Some One <someone@vm40.net>\n\n</p>messageContent";

        const contentWithSeparator = addSeparator(
            messageContent,
            previousMessage,
            MessageCreationModes.FORWARD,
            true,
            vueI18n
        );
        expect(contentWithSeparator).toEqual(expectedContent);
    });

    test("for Reply without userPrefTextOnly", async () => {
        const expectedContent =
            '<br><div id="data-bm-forward-separator"><p>On ' +
            previousMessage.date +
            `, Some One <someone@vm40.net> wrote:<br></p><blockquote style="margin-left: 1rem; padding-left: 1rem; border-left: 2px solid black;">` +
            messageContent +
            "</blockquote></div>";

        const contentWithSeparator = addSeparator(
            messageContent,
            previousMessage,
            MessageCreationModes.REPLY,
            false,
            vueI18n
        );
        expect(contentWithSeparator).toEqual(expectedContent);
    });
});

describe("compute To and Cc recipients when replying", () => {
    beforeEach(() => {
        previousMessage.headers = [];
    });

    const currentIdentity = { email: "jdoe@vm40.net", displayname: "John Doe" };
    const otherRecipients = ["azerty@keyboard.com", "memory@ram.net", "pixel@lcd.org"];
    const otherRecipientsWithDn = otherRecipients.map(address => ({ address, dn: "" }));

    test("Reply and no header", () => {
        const to = computeToRecipients(MessageCreationModes.REPLY, previousMessage, currentIdentity);
        expect(to).toEqual([previousMessageFrom]);

        const cc = computeCcRecipients(MessageCreationModes.REPLY, previousMessage);
        expect(cc).toEqual([]);
    });

    test("Reply and Mail-Followup-To header", () => {
        previousMessage.headers = [{ name: MessageHeader.MAIL_FOLLOWUP_TO, values: otherRecipients }];

        const to = computeToRecipients(MessageCreationModes.REPLY, previousMessage, currentIdentity);
        expect(to).toEqual([previousMessageFrom]);

        const cc = computeCcRecipients(MessageCreationModes.REPLY, previousMessage);
        expect(cc).toEqual([]);
    });

    test("Reply and Mail-Reply-To header", () => {
        previousMessage.headers = [{ name: MessageHeader.MAIL_REPLY_TO, values: otherRecipients }];

        const to = computeToRecipients(MessageCreationModes.REPLY, previousMessage, currentIdentity);
        expect(to).toEqual([{ address: "azerty@keyboard.com", dn: "" }]);

        const cc = computeCcRecipients(MessageCreationModes.REPLY, previousMessage);
        expect(cc).toEqual([]);
    });

    test("Reply and Reply-To header", () => {
        previousMessage.headers = [{ name: MessageHeader.REPLY_TO, values: otherRecipients }];

        const to = computeToRecipients(MessageCreationModes.REPLY, previousMessage, currentIdentity);
        expect(to).toEqual([{ address: "azerty@keyboard.com", dn: "" }]);

        const cc = computeCcRecipients(MessageCreationModes.REPLY, previousMessage);
        expect(cc).toEqual([]);
    });

    const previousToWithoutMe = previousMessageTo.filter(to => to.address !== currentIdentity.email);

    test("ReplyAll and no header", () => {
        const to = computeToRecipients(MessageCreationModes.REPLY_ALL, previousMessage, currentIdentity);
        expect(to).toEqual([previousMessageFrom].concat(previousToWithoutMe));

        const cc = computeCcRecipients(MessageCreationModes.REPLY_ALL, previousMessage);
        expect(cc).toEqual(previousMessageCc);
    });

    test("ReplyAll and Mail-Followup-To header", () => {
        previousMessage.headers = [{ name: MessageHeader.MAIL_FOLLOWUP_TO, values: otherRecipients }];

        const to = computeToRecipients(MessageCreationModes.REPLY_ALL, previousMessage, currentIdentity);
        expect(to).toEqual(otherRecipientsWithDn);

        const cc = computeCcRecipients(MessageCreationModes.REPLY_ALL, previousMessage);
        expect(cc).toEqual([]);
    });

    test("ReplyAll and Mail-Reply-To header", () => {
        previousMessage.headers = [{ name: MessageHeader.MAIL_REPLY_TO, values: otherRecipients }];

        const to = computeToRecipients(MessageCreationModes.REPLY_ALL, previousMessage, currentIdentity);
        expect(to).toEqual(otherRecipientsWithDn);

        const cc = computeCcRecipients(MessageCreationModes.REPLY_ALL, previousMessage);
        expect(cc).toEqual(previousMessageCc);
    });

    test("ReplyAll and Reply-To header", () => {
        previousMessage.headers = [{ name: MessageHeader.REPLY_TO, values: otherRecipients }];

        const to = computeToRecipients(MessageCreationModes.REPLY_ALL, previousMessage, currentIdentity);
        expect(to).toEqual(otherRecipientsWithDn);

        const cc = computeCcRecipients(MessageCreationModes.REPLY_ALL, previousMessage);
        expect(cc).toEqual(previousMessageCc);
    });

    test("Remove 'From' from recipients", () => {
        previousMessage.to.push(currentIdentity);
        const to = computeToRecipients(MessageCreationModes.REPLY_ALL, previousMessage, currentIdentity);
        expect(to.findIndex(recipient => recipient.address === currentIdentity.email)).toBe(-1);
    });
});

describe("compute From when replying or forwarding", () => {
    const defaultId = { email: "default@mail.org", name: "default", isDefault: true, displayname: "default" };
    const other = { email: "contact@mail.org", name: "contacts", isDefault: false, displayname: "Contact" };
    const anotherAgain = { email: "rh@mail.org", name: "rh", isDefault: false, displayname: "RH" };
    const identities = [defaultId, other, anotherAgain];

    const currentMailbox = { address: "rh@mail.org" };

    test("only one identity", () => {
        const message = { to: [{ address: "contact@mail.org" }], cc: [] };
        const from = computeIdentityForReplyOrForward(message, identities, currentMailbox);
        expect(from.email).toBe("contact@mail.org");
    });

    test("priorize default identity if multiple found", () => {
        const message = {
            to: [{ address: "contact@mail.org" }, { address: "default@mail.org" }],
            cc: [{ address: "rh@mail.org" }]
        };
        const defaultIdentity = { email: "plop@plop.plop", displayname: "Plop Plop" };
        const from = computeIdentityForReplyOrForward(message, identities, currentMailbox, defaultIdentity);
        expect(from.email).toBe("plop@plop.plop");
    });

    test("priorize 'to' recipient if multiple found (and no default)", () => {
        const message = { to: [{ address: "contact@mail.org" }], cc: [{ address: "rh@mail.org" }] };
        const from = computeIdentityForReplyOrForward(message, identities, currentMailbox);
        expect(from.email).toBe("contact@mail.org");
    });

    test("if no identities found in recipients, take current mailbox one", () => {
        const message = { to: [], cc: [] };
        const from = computeIdentityForReplyOrForward(message, identities, currentMailbox);
        expect(from.email).toBe("rh@mail.org");
    });

    test("if no identities found in recipients, and mailbox dont match any identity, use default", () => {
        let message = { to: [], cc: [] };
        const defaultIdentity = { email: "plop@plop.plop", displayname: "Plop Plop" };
        let from = computeIdentityForReplyOrForward(message, identities, {}, defaultIdentity);
        expect(from.email).toBe("plop@plop.plop");

        message = { to: [{ address: "blabla@mail.com" }], cc: [{ address: "moreblabla@mail.org" }] };
        from = computeIdentityForReplyOrForward(
            message,
            identities,
            {
                address: "cantfindme@mail.org",
                dn: "Cant find me"
            },
            defaultIdentity
        );
        expect(from.email).toBe("plop@plop.plop");
    });
});

describe("find identity given a current mailbox, or fallback on default", () => {
    const defaultIdentity = { email: "default@mail.org", name: "default", isDefault: true, displayname: "default" };
    const other = { email: "contact@mail.org", name: "contacts", isDefault: false, displayname: "Contact" };

    const identities = [other, defaultIdentity];
    const currentMailbox = { dn: other.displayname, address: other.email };

    test("set identity according to currentMailbox", () => {
        const identity = findIdentityFromMailbox(currentMailbox, identities, defaultIdentity);
        expect(identity).toMatchObject(other);
    });

    test("set default identity if no identity matches currentMailbox", () => {
        const identity = findIdentityFromMailbox(
            { dn: "unknown", address: "un@known.org" },
            identities,
            defaultIdentity
        );
        expect(identity).toMatchObject(defaultIdentity);
    });
});
