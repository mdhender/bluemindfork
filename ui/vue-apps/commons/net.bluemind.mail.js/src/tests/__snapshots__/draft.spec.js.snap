// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Compute reply / forward quoted previous message for Forward with userPrefTextOnly false 1`] = `
"<br><div id=\\"data-bm-forward-separator\\"><p style=\\"color: purple;\\">
            ---- Original Message ----<br>
            From: Some One &lt;someone@vm40.net&gt;<br>
            Subject: <i>undefined</i><br>
            To: John Doe &lt;jdoe@vm40.net&gt;, Toto Matic &lt;tmatic@vm40.net&gt;, Georges Abitbol &lt;gabitbol@vm40.net&gt;<br>
            Cc: Another John Doe &lt;anotherjdoe@vm40.net&gt;, Another Toto Matic &lt;anothertmatic@vm40.net&gt;, Another Georges Abitbol &lt;anothergabitbol@vm40.net&gt;<br>
            Date: Wed Dec 12 2012 00:00:00 GMT+0000 (heure moyenne de Greenwich)
        </p>My message content.</div>"
`;

exports[`Compute reply / forward quoted previous message for Forward with userPrefTextOnly true 1`] = `
"
---- Original Message ----
From: Some One <someone@vm40.net>
Subject: undefined
To: John Doe <jdoe@vm40.net>, Toto Matic <tmatic@vm40.net>, Georges Abitbol <gabitbol@vm40.net>
Cc: Another John Doe <anotherjdoe@vm40.net>, Another Toto Matic <anothertmatic@vm40.net>, Another Georges Abitbol <anothergabitbol@vm40.net>
Date: Wed Dec 12 2012 00:00:00 GMT+0000 (heure moyenne de Greenwich)My message content."
`;

exports[`Compute reply / forward quoted previous message for Reply and ReplyAll with userPrefTextOnly false 1`] = `"<br><div id=\\"data-bm-forward-separator\\"><p>On Wed Dec 12 2012 00:00:00 GMT+0000 (heure moyenne de Greenwich), Some One &lt;someone@vm40.net&gt; wrote:</p><blockquote style=\\"margin-left: 1rem; padding-left: 1rem; border-left: 2px solid black; border-left-color: var(--neutral-fg-lo1);\\">My message content.</blockquote></div>"`;

exports[`Compute reply / forward quoted previous message for Reply and ReplyAll with userPrefTextOnly false 2`] = `"<br><div id=\\"data-bm-forward-separator\\"><p>On Wed Dec 12 2012 00:00:00 GMT+0000 (heure moyenne de Greenwich), Some One &lt;someone@vm40.net&gt; wrote:</p><blockquote style=\\"margin-left: 1rem; padding-left: 1rem; border-left: 2px solid black; border-left-color: var(--neutral-fg-lo1);\\">My message content.</blockquote></div>"`;

exports[`Compute reply / forward quoted previous message for Reply and ReplyAll with userPrefTextOnly true 1`] = `
"
On Wed Dec 12 2012 00:00:00 GMT+0000 (heure moyenne de Greenwich), Some One <someone@vm40.net> wrote:


> My message content."
`;

exports[`Compute reply / forward quoted previous message for Reply and ReplyAll with userPrefTextOnly true 2`] = `
"
On Wed Dec 12 2012 00:00:00 GMT+0000 (heure moyenne de Greenwich), Some One <someone@vm40.net> wrote:


> My message content."
`;
